### ISSUE TITLE
flex attention should fuse gating in its epilogue

### ISSUE THREAD

**Issue by v0i0** [pytorch/pytorch#169435]
### 🚀 The feature, motivation and pitch

it would be nice if it did this fusion. example code:
```Py
# /// script
# dependencies = [
#   "torch",
#   "numpy",
# ]
# ///
import os
os.environ["TORCH_LOGS"] = "output_code"
import torch
def noop(score, b, h, q_idx, kv_idx):
    return score
from torch.nn.attention.flex_attention import flex_attention
query = torch.randn(1, 1, 100, 128, dtype=torch.bfloat16, device='cuda', requires_grad=False)
key = torch.randn(1, 1, 100, 128, dtype=torch.bfloat16, device='cuda', requires_grad=False)
value = torch.randn(1, 1, 100, 128, dtype=torch.bfloat16, device='cuda', requires_grad=False)
gate = torch.randn(1, 1, 100, 128, dtype=torch.bfloat16, device='cuda', requires_grad=False)

@torch.compile(mode="max-autotune")
def fn(q, k, v, g):
    return flex_attention(query, key, value, score_mod=noop) * torch.sigmoid(g)

fn(query, key, value, gate)
```

result: https://gist.github.com/v0i0/6612c12c7c531d907341ce6015396ce2

cc @chauhang @penguinwu @Chillee @drisspg @yanboliang @BoyuanFeng

**Comment by howardzhang-cv** [comment]
It seems that the epilogue fusion only occurs when we use the TRITON_DECODE backend for flex attention. In that backend, to optimize GPU usage, we need to do a reduction operation at the end, which cannot be fused with pointwise operations at the epilogue. To fix, you can either: a) use a longer sequence length, or b) force using the regular TRITON backend.

I'm attaching the output log so you can verify that the fusion is indeed happening when using the regular TRITON backend. You can see that op10 (the Triton template buffer for the flex attention kernel) is fused with op 11 (the sigmoid gating op).

[flex_gating_output_log.txt](https://github.com/user-attachments/files/24573972/flex_gating_output_log.txt)
